src/python/zquantum/core/wip/circuit/_circuit_test.py

```diff
@@ -0,0 +1,743 @@
+    RZ(2),
+    RZ(3),
+    RZ(0, sympy.Symbol("gamma")),
+    RZ(1, sympy.Symbol("gamma")),
+    RZ(2, sympy.Symbol("gamma")),
+    RZ(3, sympy.Symbol("gamma")),
+    RZ(0, 0),
+    RZ(1, 0),
+    RZ(2, 0),
+    RZ(3, 0),
+    RZ(0, 0.5),
+    RZ(1, 0.5),
+    RZ(2, 0.5),
+    RZ(3, 0.5),
+    RZ(0, np.pi / 2),
+    RZ(1, np.pi / 2),
+    RZ(2, np.pi / 2),
+    RZ(3, np.pi / 2),
+    RZ(0, np.pi),
+    RZ(1, np.pi),
+    RZ(2, np.pi),
+    RZ(3, np.pi),
+    RZ(0, 2 * np.pi),
+    RZ(1, 2 * np.pi),
+    RZ(2, 2 * np.pi),
+    RZ(3, 2 * np.pi),
+    CNOT(0, 1),
+    CNOT(0, 9),
+    CustomParameterizedGate,
+]
+
+CIRCUITS = [
+    Circuit(gates=[]),
+    Circuit(gates=[X(0)]),
+    Circuit(gates=[X(1)]),
+    Circuit(gates=[X(0), X(1)]),
+    Circuit(
+        gates=[
+            H(0),
+            CNOT(0, 1),
+            RX(0),
+            CNOT(0, 1),
+            H(0),
+        ]
+    ),
+    Circuit(gates=[CustomParameterizedGate]),
+    Circuit(
+        gates=[
+            RX(0),
+            RY(0),
+            RZ(0),
+            CustomParameterizedGate,
+        ]
+    ),
+    Circuit(gates=[I(0) for _ in range(100)]),
+    Circuit(gates=[random.choice(RandomGateList) for _ in range(100)]),
+    Circuit(gates=[random.choice(RandomGateList) for _ in range(1000)]),
+    Circuit(gates=[random.choice(RandomGateList) for _ in range(10000)]),
+]
+
+
+#### __init__ ####
+@pytest.mark.parametrize(
+    "gates",
+    [
+        [],
+        [X(0)],
+        [H(0)],
+        [X(0), X(0)],
+        [CNOT(0, 1)],
+        [X(0), X(0), CNOT(0, 1), X(0)],
+        [random.choice(RandomGateList) for _ in range(100)],
+    ],
+)
+def test_creating_circuit_has_correct_gates(gates):
+    """The Circuit class should have the correct gates that are passed in"""
+    # When
+    circuit = Circuit(gates=gates)
+    # Then
+    assert circuit.gates == gates
+
+
+def test_appending_to_circuit_works():
+    """The Circuit class should have the correct gates that are passed in"""
+    # Given
+    expected_circuit = Circuit(gates=[H(0), CNOT(0, 1)])
+    # When
+    circuit = Circuit(gates=[])
+    circuit.gates.append(H(0))
+    circuit.gates.append(CNOT(0, 1))
+    # Then
+    assert circuit.gates == expected_circuit.gates
+    assert circuit.qubits == expected_circuit.qubits
+
+
+#### qubits ####
+@pytest.mark.parametrize(
+    "gates, qubits",
+    [
+        ([], tuple()),
+        ([X(0)], (0,)),
+        ([X(1)], (1,)),
+        (
+            [X(0), X(1)],
+            (
+                0,
+                1,
+            ),
+        ),
+        ([CNOT(0, 1)], (0, 1)),
+        ([X(0), X(0), CNOT(0, 1), X(0)], (0, 1)),
+    ],
+)
+def test_creating_circuit_has_correct_qubits(gates, qubits):
+    """The Circuit class should have the correct qubits based on the gates that are passed in"""
+    # When
+    circuit = Circuit(gates=gates)
+    # Then
+    assert circuit.qubits == qubits
+
+
+def test_creating_circuit_has_correct_qubits_with_gaps():
```

Comment:
No, I actually think this is correct. This does though kind of highlight the difference between a quantum "program" or logical circuit vs the physical quantum circuit. Perhaps something to think about for the future

Reply:
Though, couldn't this be combined with the above test, just a special case?

Reply:
I think the confusion is related to ambiguity of the property name `qubits`. An idea I'm having now is to rename it to `active_qubits` or `used_qubits`, but I'd be more confident if I knew use cases for it. The whole `Circuit` will be revamped in subsequent PRs, I hope to hear more about the use cases from potential users.
 @dexter2206

Reply:
If @simonwa7 says it's fine, I'm good.
I don't think going with the `active_qubits` is a good idea, as it might cause more confusion than help. Though haven't thought about it carefully .